# Demo: Dataviews on OSIsoft Cloud Service (OCS) 

Dataviews allows data scientists to create and use table views which can be vary from simple (e.g extracting values from a single sensor) to very elaborate by pulling data from multiple sources and combining them. The result of a Dataview is a standard CSV (comma separated value) table ready-to-use  

The dataset used below is from the Deschutes Brewery, with data coming from fermentor vessels. 

This notebook shows the steps involved in creating and using Dataviews. 

## Imports 

In [1]:
# In a CMD window, execute: pip install requests-futures plotly
# 
# For interaction with OCS
from ocs_datascience import OCSClient
# For parallel HTTP requests
from concurrent.futures import ThreadPoolExecutor
from requests_futures.sessions import FuturesSession
import requests
# Pandas dataframe to manipulate table data
import pandas as pd
# Utilities from Python standard library 
import configparser
import datetime
import io
import json

## Content of file `config.ini`

    ; IMPORTANT: those values are provided by OSIsoft, DO NOT CHANGE
    [Configurations]
    Namespace = fermenter_vessels

    [Access]
    Resource = https://dat-b.osisoft.com
    Tenant = d7847614-2e4a-4c1e-812b-e8de5fd06a0f
    ApiVersion = v1-preview

    [Credentials]
    ClientId = ec9d0232-fc61-4ee7-8316-6997954ad40c
    ClientSecret = OmY31XkODlimHSR5gDZqqE5PT8HrUm3liDsIjwNc5VQ=
    
## Read in configuration file and create OCS client object

In [2]:
config = configparser.ConfigParser()
config.read('config.ini')

ocs_client = OCSClient(config.get('Access', 'ApiVersion'),config.get('Access', 'Tenant'), config.get('Access', 'Resource'), 
                     config.get('Credentials', 'ClientId'), config.get('Credentials', 'ClientSecret'))

namespace_id = config.get('Configurations', 'Namespace')

## Get an the autorization header with bearer token for access to OCS API 

In [3]:
headers = ocs_client.authorization_headers()
headers

Authorization: <Response [200]> {"access_token":"eyJhbGciOiJSUzI1NiIsImtpZCI6IjJDQjI4MzFEREJFRDc1NzAyM0NCMTM5OUVBRjRDMjkxQzE3MkQ5RjQiLCJ0eXAiOiJKV1QiLCJ4NXQiOiJMTEtESGR2dGRYQWp5eE9aNnZUQ2tjRnkyZlEifQ.eyJuYmYiOjE1NTQwMzkwNjksImV4cCI6MTU1NDA0MjY2OSwiaXNzIjoiaHR0cHM6Ly9kYXQtYi5vc2lzb2Z0LmNvbS9pZGVudGl0eSIsImF1ZCI6WyJodHRwczovL2RhdC1iLm9zaXNvZnQuY29tL2lkZW50aXR5L3Jlc291cmNlcyIsIm9jc2FwaSJdLCJjbGllbnRfaWQiOiJhNDkxMjM1My00OWEzLTRlYzAtYTdkYS1mMGZmN2U0YjM0YmEiLCJ0aWQiOiJkNzg0NzYxNC0yZTRhLTRjMWUtODEyYi1lOGRlNWZkMDZhMGYiLCJqdGkiOiIyNzUxMWQ1ZjdiNzM1OGI5MTYyNjY4MWZiMzQyNjYzYiIsInNjb3BlIjpbIm9jc2FwaSJdfQ.X2giIZS4-tw9xiHKc8j2aEG-U6hmXOEcEW0rnYvFK9s9--4qGFQUjlZb1zpOW4KQJWuCa36uaNoRdEfl-iDCG-jqRXyyzuCMY3wcZA1oT5VPCw4WF2XLAoaRInNXjWuabshRb0CgWtqBPm1Q10zoiSc9pFQ1auP0Ms4muqwapZPvEIH8MYn4eLXrruSIXm16ruMs99RQ5ADLQtx45sbZfUKVdznd-BijmyvzpGOJcO3Dh-SdVSPUqD0tAktbyUuSddYBzyMur-HHrx7d0ojXrqv77jq5BQp9gqSpJLzmaH7eZ0def-pDWiw-_xWq3TgxJSAVDqx4iD4PciYHMQ2clg","expires_in":3600,"token_type":"Bearer"}


{'Authorization': 'bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IjJDQjI4MzFEREJFRDc1NzAyM0NCMTM5OUVBRjRDMjkxQzE3MkQ5RjQiLCJ0eXAiOiJKV1QiLCJ4NXQiOiJMTEtESGR2dGRYQWp5eE9aNnZUQ2tjRnkyZlEifQ.eyJuYmYiOjE1NTQwMzkwNjksImV4cCI6MTU1NDA0MjY2OSwiaXNzIjoiaHR0cHM6Ly9kYXQtYi5vc2lzb2Z0LmNvbS9pZGVudGl0eSIsImF1ZCI6WyJodHRwczovL2RhdC1iLm9zaXNvZnQuY29tL2lkZW50aXR5L3Jlc291cmNlcyIsIm9jc2FwaSJdLCJjbGllbnRfaWQiOiJhNDkxMjM1My00OWEzLTRlYzAtYTdkYS1mMGZmN2U0YjM0YmEiLCJ0aWQiOiJkNzg0NzYxNC0yZTRhLTRjMWUtODEyYi1lOGRlNWZkMDZhMGYiLCJqdGkiOiIyNzUxMWQ1ZjdiNzM1OGI5MTYyNjY4MWZiMzQyNjYzYiIsInNjb3BlIjpbIm9jc2FwaSJdfQ.X2giIZS4-tw9xiHKc8j2aEG-U6hmXOEcEW0rnYvFK9s9--4qGFQUjlZb1zpOW4KQJWuCa36uaNoRdEfl-iDCG-jqRXyyzuCMY3wcZA1oT5VPCw4WF2XLAoaRInNXjWuabshRb0CgWtqBPm1Q10zoiSc9pFQ1auP0Ms4muqwapZPvEIH8MYn4eLXrruSIXm16ruMs99RQ5ADLQtx45sbZfUKVdznd-BijmyvzpGOJcO3Dh-SdVSPUqD0tAktbyUuSddYBzyMur-HHrx7d0ojXrqv77jq5BQp9gqSpJLzmaH7eZ0def-pDWiw-_xWq3TgxJSAVDqx4iD4PciYHMQ2clg',
 'Content-type': 'application/json',
 'Accept': 'text/plain',
 'Request-Timeout

### URL to access `fermenter_vessels` namespace and its dataviews 

In [4]:
# Endpoint for dataview access
namespace_url = ocs_client.namespace_url(namespace_id)  
dataview_url = namespace_url + '/dataviews/'
namespace_url

'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels'

### Example of typical interaction with OCS: request and display all data streams of Fermentor Vessel 31

#### 1) Build a stream query URL

In [5]:
streams_url = namespace_url + '/Streams?query=name:*FV31*'
print('Stream Query URL:', streams_url)

Stream Query URL: https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/Streams?query=name:*FV31*


#### 2) Make an HTTP GET request using URL and authorization header

In [6]:
fv31_streams = requests.get(streams_url, headers=headers)

#### 3) Verify that request status code indicates success (should be 200 for GET) 

In [7]:
fv31_streams.status_code

200

#### 4) Display result in JSON format (pretty-print)

In [8]:
print(json.dumps(fv31_streams.json(), indent=4))

[
    {
        "TypeId": "PIFloat32",
        "Id": "PI_acad-pida-vm0_2592",
        "Name": "acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/OUT.CV",
        "Description": "FV31 Bottom Temperature Control Output",
        "InterpolationMode": 0,
        "ExtrapolationMode": 2
    },
    {
        "TypeId": "PIFloat32",
        "Id": "PI_acad-pida-vm0_2639",
        "Name": "acsbrew.BREWERY.B2_CL_C2_FV31_PIC1362/SP.CV",
        "Description": "FV31 Pressure Control Setpoint",
        "InterpolationMode": 0,
        "ExtrapolationMode": 2
    },
    {
        "TypeId": "PIDigital",
        "Id": "PI_acad-pida-vm0_2968",
        "Name": "acsbrew.BREWERY.B2_CL_C2_FV31/YEAST.CV",
        "Description": "FV31 Yeast",
        "InterpolationMode": 1,
        "ExtrapolationMode": 2
    },
    {
        "TypeId": "PIDigital",
        "Id": "PI_acad-pida-vm0_2598",
        "Name": "acsbrew.BREWERY.B2_CL_C2_FV31/BRAND.CV",
        "Description": "FV31 Brand",
        "InterpolationMode": 1,
        "Ext

## Extract and display only stream names 

In [9]:
# Iterate over each stream and extract the field of interest with j['field'] syntax
streams = [stream['Name'] for stream in fv31_streams.json()]
streams

['acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/OUT.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31_PIC1362/SP.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/YEAST.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/BRAND.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/STATUS.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/SP.CV',
 'acsbrew.BREWERY.FV31.ADF2',
 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/OUT.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31_PIC1362/PV.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/PV.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/HARVEST.CV',
 'acsbrew.BREWERY.B2_%@Area%_FV31.OEE.Performance',
 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/SP.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/DcrsFvFullPlato',
 'acsbrew.BREWERY.B2_CL_C2_FV31/PCD.CV',
 'acsbrew.BREWERY.B2_%@Area%_FV31.OEE.Quality',
 'acsbrew.BREWERY.B2_%@Area%_FV31.OEE.Availability',
 'acsbrew.BREWERY.FV31.Fermentation ID.194fa814-869f-5f35-3501-0b9198ac52e1',
 'acsbrew.BREWERY.B2_CL_C2_FV31/YEASTGEN.CV',
 'acsbrew.BREWERY.B2_CL_C2_FV31/PULLS

## Mapping of the streams of interest to a dataview column 

| Stream Name | DV Column Name | Description | 
|-------------|----------------|-------------|
| acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV | `Volume` | Vessel Volume 
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/PV.CV | `Top TIC PV` | Vessel Bottom Temperture Indicator Controller Process Value
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/OUT.CV | `Top TIC OUT` | Vessel Top Temperature Indicator Controller Output
| acsbrew.BREWERY.B2_CL_C2_FV31/Plato | `Plato` | The specific gravity of the vessel in plato
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/PV.CV | `Middle TIC PV` | Vessel Middle Temperture Indicator Controller Process 
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/OUT.CV | `Middle TIC OUT` | Vessel Middle Temperature Indicator Controller Output
| acsbrew.BREWERY.B2_CL_C2_FV31/DcrsFvFullPlato | `FV Full Plato` | The specific gravity of the vessel in plato at the end of filling
| acsbrew.BREWERY.FV31.Fermentation ID.194fa814-869f-5f35-3501-0b9198ac52e1 | `Fermentation ID` | Unique ID for fermentation batch 
| acsbrew.BREWERY.B2_CL_C2_FV31/BRAND.CV | `Brand` | Vessel Brand
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/PV.CV | `Bottom TIC PV` | Vessel Bottom Temperture Indicator Controller Process Value
Value
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/OUT.CV |`Bottom TIC OUT` | Vessel Bottom Temperature Indicator Controller Output
| acsbrew.BREWERY.FV31.ADF2 | `ADF` | Apparent Degree of Fermentation 
| acsbrew.BREWERY.B2_CL_C2_FV31/STATUS.CV | `Status` | * Vessel Status 

### Here is what Dataview definition of fermenter vessel 31 looks like

In [10]:
dv_id, dv_def = ocs_client.fermenter_dataview_def(31)
dv_def

{'Id': 'DV_FV31',
 'Name': 'DV_FV31',
 'Description': 'Fermentor 31 DV',
 'Queries': [{'Id': 'Fermentor',
   'Query': {'Resource': 'Streams',
    'Field': 'Name',
    'Value': 'FV31',
    'Operator': 'Contains'}}],
 'GroupRules': [],
 'Mappings': {'Columns': [{'Name': 'Timestamp',
    'MappingRule': {'PropertyPaths': ['Timestamp']},
    'IsKey': True,
    'DataType': 'DateTime'},
   {'Name': 'Volume',
    'MappingRule': {'PropertyPaths': ['Value'],
     'ItemIdentifier': {'Resource': 'Streams',
      'Field': 'Name',
      'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV',
      'Operator': 'Equals'}}},
   {'Name': 'Top TIC PV',
    'MappingRule': {'PropertyPaths': ['Value'],
     'ItemIdentifier': {'Resource': 'Streams',
      'Field': 'Name',
      'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/PV.CV',
      'Operator': 'Equals'}}},
   {'Name': 'Top TIC OUT',
    'MappingRule': {'PropertyPaths': ['Value'],
     'ItemIdentifier': {'Resource': 'Streams',
      'Field': 'Name',
  

### Let's extract all the columns definition 

In [11]:
dv_def['Mappings']['Columns']

[{'Name': 'Timestamp',
  'MappingRule': {'PropertyPaths': ['Timestamp']},
  'IsKey': True,
  'DataType': 'DateTime'},
 {'Name': 'Volume',
  'MappingRule': {'PropertyPaths': ['Value'],
   'ItemIdentifier': {'Resource': 'Streams',
    'Field': 'Name',
    'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV',
    'Operator': 'Equals'}}},
 {'Name': 'Top TIC PV',
  'MappingRule': {'PropertyPaths': ['Value'],
   'ItemIdentifier': {'Resource': 'Streams',
    'Field': 'Name',
    'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/PV.CV',
    'Operator': 'Equals'}}},
 {'Name': 'Top TIC OUT',
  'MappingRule': {'PropertyPaths': ['Value'],
   'ItemIdentifier': {'Resource': 'Streams',
    'Field': 'Name',
    'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/OUT.CV',
    'Operator': 'Equals'}}},
 {'Name': 'Plato',
  'MappingRule': {'PropertyPaths': ['Value'],
   'ItemIdentifier': {'Resource': 'Streams',
    'Field': 'Name',
    'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31/Plato',
    'Operator': 'Equals'}

### Iterate over the list of Dataview columns and print a table: Stream ID, Column Name

In [12]:
columns = dv_def['Mappings']['Columns']
# Remove first column which is the Timestamp
columns = columns[1:]
# For each column item c which has the following form:
#  {'Name': 'Volume',
#   'MappingRule': {
#      'PropertyPaths': ['Value'],
#      'ItemIdentifier': {
#         'Resource': 'Streams',
#         'Field': 'Name',
#         'Value': 'acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV',
#         'Operator': 'Equals'}}}
# 
# extract the stream ID (acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV above) and column name (Volume above)
streams_and_columns = [ (c['MappingRule']['ItemIdentifier']['Value'], c['Name']) for c in columns]
for stream, column in streams_and_columns:
    print(f'{stream:<75}        {column}')

acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV                                         Volume
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/PV.CV                                       Top TIC PV
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/OUT.CV                                      Top TIC OUT
acsbrew.BREWERY.B2_CL_C2_FV31/Plato                                                Plato
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/PV.CV                                       Middle TIC PV
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/OUT.CV                                      Middle TIC OUT
acsbrew.BREWERY.B2_CL_C2_FV31/DcrsFvFullPlato                                      FV Full Plato
acsbrew.BREWERY.FV31.Fermentation ID.194fa814-869f-5f35-3501-0b9198ac52e1          Fermentation ID
acsbrew.BREWERY.B2_CL_C2_FV31/BRAND.CV                                             Brand
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/PV.CV                                       Bottom TIC PV
acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/OUT.CV          

### Check that the list above match the table of mapping given before (reproduced below)

| Stream Id | DV Column Name | Description | 
|-------------|----------------|-------------|
| acsbrew.BREWERY.B2_CL_C2_FV31_LT1360/PV.CV | `Volume` | Vessel Volume 
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/PV.CV | `Top TIC PV` | Vessel Bottom Temperture Indicator Controller Process Value
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360C/OUT.CV | `Top TIC OUT` | Vessel Top Temperature Indicator Controller Output
| acsbrew.BREWERY.B2_CL_C2_FV31/Plato | `Plato` | The specific gravity of the vessel in plato
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/PV.CV | `Middle TIC PV` | Vessel Middle Temperture Indicator Controller Process 
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360B/OUT.CV | `Middle TIC OUT` | Vessel Middle Temperature Indicator Controller Output
| acsbrew.BREWERY.B2_CL_C2_FV31/DcrsFvFullPlato | `FV Full Plato` | The specific gravity of the vessel in plato at the end of filling
| acsbrew.BREWERY.FV31.Fermentation ID.194fa814-869f-5f35-3501-0b9198ac52e1 | `Fermentation ID` | Unique ID for fermentation batch 
| acsbrew.BREWERY.B2_CL_C2_FV31/BRAND.CV | `Brand` | Vessel Brand
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/PV.CV | `Bottom TIC PV` | Vessel Bottom Temperture Indicator Controller Process Value
| acsbrew.BREWERY.B2_CL_C2_FV31_TIC1360A/OUT.CV |`Bottom TIC OUT` | Vessel Bottom Temperature Indicator Controller Output
| acsbrew.BREWERY.FV31.ADF2 | `ADF` | Apparent Degree of Fermentation 
| acsbrew.BREWERY.B2_CL_C2_FV31/STATUS.CV | `Status` | * Vessel Status 

## Creation of the Dataviews, for fermenters 31 up to 36

* Status 201 from POST request indicates success
* Status 401 indicates unauthorized (try refreshing authorization header)
* Status 409 when a Dataview with same Id already exists (go to last cell of this notebook to perform a clean up)

In [13]:
dataviews = []
for fv_id in range(31, 37):  
    dataview_id, dataview_def = ocs_client.fermenter_dataview_def(fv_id)
    dataviews.append(dataview_id)
    response = requests.post(dataview_url + dataview_id, headers=headers, json=dataview_def)
    print('Status:', response.status_code, 'DV ID:', dataview_id)

Status: 400 DV ID: DV_FV31
Status: 400 DV ID: DV_FV32
Status: 400 DV ID: DV_FV33
Status: 400 DV ID: DV_FV34
Status: 400 DV ID: DV_FV35
Status: 400 DV ID: DV_FV36


## List of Dataviews URLs 

In [19]:
# We want 20 days of data worth at 1 minute interval, for fermenter 31 up to 36
dataviews_url = [dataview_url + '%s/preview/interpolated?form=csvh&maxcount=100000' % d for d in dataviews]
dataviews_url

['https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/dataviews/DV_FV31/preview/interpolated?form=csvh&maxcount=100000',
 'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/dataviews/DV_FV32/preview/interpolated?form=csvh&maxcount=100000',
 'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/dataviews/DV_FV33/preview/interpolated?form=csvh&maxcount=100000',
 'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/dataviews/DV_FV34/preview/interpolated?form=csvh&maxcount=100000',
 'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5fd06a0f/Namespaces/fermenter_vessels/dataviews/DV_FV35/preview/interpolated?form=csvh&maxcount=100000',
 'https://dat-b.osisoft.com/api/v1-preview/Tenants/d7847614-2e4a-4c1e-812b-e8de5

## From a list of dataviews, gather them in parallel and return a single dataframe

In [20]:
def postproc(reply):
    s = io.StringIO(reply)
    # 255 Bad Input, 307 Bad, 313 Comm Fail, 246 I/O Timeout, 
    # print('Line 1:', s.readline())
    # print('Line 2:', s.readline())
    s.close()
    return reply 

# Request in parallel all the dataviews, return the concatenated dataframe
def get_ocs_dataframe(dataviews, headers, workers=8):
    ti = datetime.datetime.now()
    session = FuturesSession(executor=ThreadPoolExecutor(max_workers=workers))
    rs = [session.get(u, headers=headers, params={'count': '10000'}) for u in dataviews]
    resps = [r.result() for r in rs]
    print('Requests completed in', datetime.datetime.now() - ti) 
    print(resps)
    dfs = [pd.read_csv(io.StringIO(postproc(r.text)), parse_dates=['Timestamp']) for r in resps]
    return(pd.concat(dfs, sort=True))

## Get dataframe and time it (up to a minute, be patient)

All responses should have HTTP code [200] to indicat everything is ok 

In [21]:
t0 = datetime.datetime.now()
df = get_ocs_dataframe(dataviews_url, headers)
print('Dataframe obtained in', datetime.datetime.now() - t0) 
df.info

Requests completed in 0:00:00.250895
[<Response [404]>, <Response [404]>, <Response [404]>, <Response [404]>, <Response [404]>, <Response [404]>]


ValueError: 'Timestamp' is not in list

### Note the that resulting dataframe has almost 182K rows

## Save data into CSV file locally in directory FlashcARD

In [ ]:
df.to_csv('beer_ocs_v3.csv')

---
## Clicking this [link](./beer_20_days.csv) opens up a CSV browser with the CSV above
---

### List of column names with their type

Note that the `Timestamp` column (a new column on top of the 16 ones of a Fermentor) has the correct datetime panda data type for timestamps

In [ ]:
for i, c in enumerate(df.columns, 1):
    print('%2d' % i, c, '((( type:', df[c].dtype, ')))')

In [ ]:
for f in df.Fermentation_ID.unique():
    print(f, isinstance(f, str))

### Prepare ADF curve plots over time 

In [ ]:
import plotly.graph_objs as go

figs = []
data = []
for f in df.Fermentation_ID.unique():
    trace = go.Scattergl(x = df[df.Fermentation_ID == f]['_time'], y = df[df.Fermentation_ID == f]['ADF'], mode='lines+markers', name=str(f))
    figs.append(go.FigureWidget(data=[trace]))
    data.append(trace)

### Add a range slider 

With a few time range selectors: 8 hours, 1 day and everything 

Note: range slider is grey now because of an incompatibility with ScatterGL: https://github.com/plotly/plotly.js/issues/2627

In [ ]:
layout = dict(
    title='Brewing ADF with time range slider',
    xaxis=dict(
        rangeselector=dict(
            buttons=list([
                dict(count=8,
                     label='8h',
                     step='hour',
                     stepmode='backward'),
                dict(count=1,
                     label='1d',
                     step='day',
                     stepmode='backward'),
                dict(step='all')
            ])
        ),
        rangeslider=dict(
            visible = True
        ),
        type='date'
    )
)
        
fig = go.FigureWidget(data=data, layout=layout)
fig

## Clean up: delete Dataviews  

* Code 204 if deletion is successful
* Code 404 if requested Dataview Id doesn't exist or already deleted

In [ ]:
for dv in dataviews_url:
    dv_url = dv[:dv.find('/preview')]
    s = requests.delete(dv_url, headers=headers)
    print(s.status_code, dv_url)

In [ ]:
full_map = [
    ('_LT', 'Volume'),
    ('C/PV.CV', 'Top TIC PV'),
    ('C/OUT.CV', 'Top TIC OUT'),
    ('/Plato', 'Plato'),
    ('B/PV.CV', 'Middle TIC PV'),
    ('B/OUT.CV', 'Middle TIC OUT'),
    ('FullPlato', 'FV Full Plato'),
    ('Fermentation', 'Fermentation ID'),
    ('BRAND', 'Brand'),
    ('A/PV.CV', 'Bottom TIC PV'),
    ('A/OUT.CV', 'Bottom TIC OUT'),
    ('STATUS', 'Status'),
]
streams, stream_tags = ocs_client.extract_streams_for_fermenter(31)
streams

In [ ]:
dv_names = [ (s, y) for (x, y) in full_map for s in streams if x in s] 
dv_names